app/spelling/.pyspelling.yml

```diff
@@ -28,7 +28,7 @@ matrix:
 - name: markdown
   expect_match: false
   sources:
-  - ./**/(*.md|!CHANGELOG.md)
```

Comment:
I was searching around for something related to wcmatch, but saw this. As I am also the author of wcmatch as well as pyspelling, I thought I'd chime in to at least let you know why this doesn't work.

You can use this group syntax unless you enable EXTGLOB: https://facelessuser.github.io/wcmatch/glob/#globextglob. Even then, there is a special syntax where each `()` must have a preceding token such as `?`, `*`, `@`, `+`, or `!` to select a group type. Documentation is covered here https://facelessuser.github.io/wcmatch/glob/#syntax.

Now, I'm not exactly sure what the aim was here, but if you have a specific question, I can most likely direct you on how you can achieve your goal.

Anyways, let me know if you are interested in exploring solutions. I see you are locked on wcmatch < 5.0 for some reason, but I assume it is related to some confusion as to how to use wcmatch, but I'm not sure. I'd love to fix issues if you feel there is an actual problem.

Reply:
Thanks so much @facelessuser - I think I got this working via https://github.com/resplendent-dev/spelling/pull/16/files - atleast I haven't found a situation it fails :) I looked over the wcmatch documentation that listed an example of using exclusions via passing a list rather than a single string. I tried this and it seems to work.

My goal was to exclude CHANGELOG.md from spell check as I've observed a fair majority of repository maintainers dislike receiving pull requests that only fix typos in the CHANGELOG.md and so I wanted to exclude it.

Reply:
> I think I got this working via https://github.com/resplendent-dev/spelling/pull/16/files

Ahh, it looks like you are sending in a nested list. I guess that works, though I guess I didn't intend it to take a list of patterns, but the API passes through okay....at least in the current implementation.

Normally I'd recommend something like this as I have `glob.SPLIT` enabled by default:

```py
>>> from wcmatch import glob
>> glob.glob('**/*.md|!**/changelog.md', flags=glob.N|glob.B|glob.S|glob.G)
['docs/src/markdown/fnmatch.md', 'docs/src/markdown/glob.md', 'docs/src/markdown/index.md', 'docs/src/markdown/license.md', 'docs/src/markdown/pathlib.md', 'docs/src/markdown/wcmatch.md', 'LICENSE.md', 'README.md']
```

Using `|` we provide two patterns. `!**/changelog.md` is applied as a negative pattern to filter out `docs/src/mardkown/changelog.md` from the positive pattern's (`**/*.md`) results.

What you are doing, which I didn't exactly intend to allow through pyspelling's API  (tough it is perfectly valid through wcmatch's API), is:

```py
>>> glob.glob(['**/*.md', '!**/changelog.md'], flags=glob.N|glob.B|glob.S|glob.G)
['docs/src/markdown/fnmatch.md', 'docs/src/markdown/glob.md', 'docs/src/markdown/index.md', 'docs/src/markdown/license.md', 'docs/src/markdown/pathlib.md', 'docs/src/markdown/wcmatch.md', 'LICENSE.md', 'README.md']
```

Essentially it is the same thing.  `glob.SPLIT` is kind of a way to represent a list through interfaces that take strings, which PySpelling is supposed to be using, though it looks like the list sneaks through just fine.

If you enable EXTGLOB (based off a real feature in Bash's glob), you can use group notation:

```py
>>> glob.glob('**/!(changelog).md', flags=glob.N|glob.B|glob.S|glob.G|glob.E)
['docs/src/markdown/fnmatch.md', 'docs/src/markdown/glob.md', 'docs/src/markdown/index.md', 'docs/src/markdown/license.md', 'docs/src/markdown/pathlib.md', 'docs/src/markdown/wcmatch.md', 'LICENSE.md', 'README.md']
```

This tells it to find anything that is not `changelog` followed by `.md`.

If you are using EXTGLOB, I usually recommend enabling MINUSNEGATE as well as it creates a conflict with the negate feature used early when the negate group is used at the start of a pattern: `!(something).md`. In this case it is difficult to tell if we are supposed to just parse `!` or `!(...)` . So by using MINUSNEGATE, `!` is now represented to by `-`. Then `-` and `!(...)` can coexist with no issues.

To enable EXTGLOB, you'd just pass in your new required flags via (here I enable EXTGLOB and MINUSNEGATE with (E and M):

```yaml
glob_flags: N|G|B|E|M
```

Obviously you can omit any flags you don't want.

Reply:
Just in case it wasn't clear though, the default value of `glob_flags` is `N|G|B` (NEGATE, GLOBSTAR, and BRACE). SPLIT cannot be turned off in PySpelling.